pkg/validation/validation.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/openshift/managed-upgrade-operator/pkg/osd_cluster_upgrader"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+// NewBuilder returns a validationBuilder object that implements the ValidationBuilder interface.
+func NewBuilder() ValidationBuilder {
+	return &validationBuilder{}
+}
+
+// Validator knows how to validate UpgradeConfig CRs.
+//go:generate mockgen -destination=mockValidation.go -package=validation github.com/openshift/managed-upgrade-operator/pkg/validation Validator
+type Validator interface {
+	IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (bool, error)
+}
+
+type validator struct{}
+
+func (v *validator) IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (bool, error) {
+	// Validate upgradeAt as RFC3339
+	_, err := time.Parse(time.RFC3339, uC.Spec.UpgradeAt)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse upgradeAt:%s during validation: %v", uC.Spec.UpgradeAt, err)
+	}
+
+	// Validate desired version.
+	dv := uC.Spec.Desired.Version
+	cv, err := osd_cluster_upgrader.GetCurrentVersion(cV)
+	if err != nil {
+		return false, fmt.Errorf("Failed to get current cluster version during validation: %v", err)
+	}
+
+	// Check for valid SemVer and convert to SemVer.
+	desiredVersion, err := semver.Parse(dv)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse desired version %s as semver: %v", dv, err)
+	}
+	currentVersion, err := semver.Parse(cv)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse current version %s as semver: %v", cv, err)
+	}
+
+	// Compare versions to ascertain if upgrade should proceed.
+	proceed := compareVersions(desiredVersion, currentVersion, logger)
+	if !proceed {
+		return false, nil
+	}
+	logger.Info(fmt.Sprintf("Desired version %s validated as greater then current version %s", desiredVersion, currentVersion))
+
+	// Validate available version is in Cincinnati.
+	desiredChannel := uC.Spec.Desired.Channel
+	clusterId, err := uuid.Parse(string(cV.Spec.ClusterID))
+	if err != nil {
+		return false, err
+	}
+	upstreamURI, err := url.Parse(string(cV.Spec.Upstream))
+	if err != nil {
+		return false, err
+	}
+
+	updates, err := cincinnati.NewClient(clusterId, nil, nil).GetUpdates(upstreamURI, runtime.GOARCH, desiredChannel, currentVersion)
+	if err != nil {
+		return false, err
+	}
+
+	var cvoUpdates []configv1.Update
+	for _, update := range updates {
+		cvoUpdates = append(cvoUpdates, configv1.Update{
+			Version: update.Version.String(),
+			Image:   update.Image,
+		})
+	}
+
+	// Check whether the desired version exists in availableUpdates
+	found := false
+	for _, v := range cvoUpdates {
+		if v.Version == dv && !v.Force {
+			found = true
+		}
+	}
+
+	if !found {
+		logger.Info(fmt.Sprintf("Failed to find the desired version %s in channel %s", desiredVersion, desiredChannel))
```

Comment:
As per https://github.com/openshift/cluster-version-operator/blob/master/pkg/cincinnati/cincinnati.go#L120 and https://github.com/openshift/cluster-version-operator/blob/master/lib/validation/validation.go there are certain validations done already by CVO. For this check regarding version check from channel, do we still need it again as part of upgrade operator? I think it would be also helpful to explore the re-sync period when CVO would perform the validation again after a change and let CVO report error if something is wrong. This is about keeping the validation checks of product via product itself if possible.

Reply:
so AFAIK, we were leaving cincinatti validation in until OCM was infront of the service and sending down the curated versions in the upgradeConfig CR. If we want to change cincinatti validation, we should define our requirements in another task.

Reply:
As discussed on Slack, this could be taken up in https://issues.redhat.com/browse/OSD-4152 further on after considering how OCM decides to handle it.. For now to keep it as it is..